File: src/main/java/com/guardanis/blank/tools/views/ToolbarLayoutBuilder.java
package com.guardanis.blank.tools.views;

import android.graphics.Color;
import android.support.annotation.Nullable;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.guardanis.blank.R;
import com.guardanis.imageloader.views.SVGImageView;

public class ToolbarLayoutBuilder {

    private Toolbar toolbar;

    public ToolbarLayoutBuilder(Toolbar toolbar){
        this(toolbar, R.layout.base__toolbar);
    }

    public ToolbarLayoutBuilder(ActionBar actionBar, int inflatedLayoutResId){
        actionBar.setDisplayShowHomeEnabled(false);
        actionBar.setDisplayShowTitleEnabled(false);
        actionBar.setDisplayHomeAsUpEnabled(false);
        actionBar.setHomeButtonEnabled(false);
        actionBar.setDisplayShowCustomEnabled(true);
        actionBar.setCustomView(inflatedLayoutResId);

        this.toolbar = (Toolbar) actionBar.getCustomView().getParent();

        applyInsetChanges();
    }

    public ToolbarLayoutBuilder(Toolbar toolbar, int inflatedLayoutResId){
        this.toolbar = toolbar;

        toolbar.removeAllViews();
        toolbar.addView(LayoutInflater.from(toolbar.getContext())
                .inflate(inflatedLayoutResId, toolbar, false));

        applyInsetChanges();
    }

    private void applyInsetChanges(){
        try{
            toolbar.setPadding(0, 0, 0, 0); // Fix for tablets
            toolbar.setContentInsetsAbsolute(0, 0);
        }
        catch(Throwable e){ e.printStackTrace(); }
    }

    public ToolbarLayoutBuilder addTitle(int textResId, View.OnClickListener onClickListener) {
        return addTitle(toolbar.getContext().getString(textResId), onClickListener);
    }

    public ToolbarLayoutBuilder addTitle(String text, View.OnClickListener onClickListener) {
        TextView textTitle = inflateTitleText();

        textTitle.setText(text);
        textTitle.setOnClickListener(onClickListener);

        addTitle(textTitle);

        return this;
    }

    public ToolbarLayoutBuilder addTitleSvg(String assetFileName, View.OnClickListener onClickListener) {
        SVGImageView svg = buildSvgImageView(assetFileName, onClickListener);
        addTitle(svg);

        return this;
    }

    public ToolbarLayoutBuilder addOptionText(String text) {
        return addOptionText(text, null);
    }

    public ToolbarLayoutBuilder addOptionText(String text, View.OnClickListener onClickListener) {
        TextView textOption = buildTextOption(text, onClickListener);

        addOption(textOption);

        return this;
    }

    public ToolbarLayoutBuilder addOptionSvg(String assetFileName, View.OnClickListener onClickListener) {
        SVGImageView svg = buildSvgImageView(assetFileName, onClickListener);
        addOption(svg);

        return this;
    }

    public ToolbarLayoutBuilder addOptionImage(int resId, View.OnClickListener onClickListener) {
        SVGImageView svg = inflateOptionImage();

        svg.setImageResource(resId);
        svg.setOnClickListener(onClickListener);

        addOption(svg);

        return this;
    }

    public SVGImageView buildSvgImageView(String assetFileName, View.OnClickListener onClickListener){
        SVGImageView svg = inflateOptionImage();
        svg.setImageAsset(assetFileName);
        svg.setOnClickListener(onClickListener);

        return svg;
    }

    public TextView buildTextOption(String text){
        return buildTextOption(text, null);
    }

    public TextView buildTextOption(String text, @Nullable View.OnClickListener onClickListener){
        TextView textView = inflateOptionText();
        textView.setText(text);
        textView.setOnClickListener(onClickListener);

        if(onClickListener != null)
            ViewHelper.setBackgroundResourceAndKeepPadding(textView, R.drawable.base__button_transparent);

        return textView;
    }

    public TextView inflateTitleText(){
        return (TextView) LayoutInflater.from(toolbar.getContext())
                .inflate(R.layout.base__toolbar_title_text, toolbar, false);
    }

    public TextView inflateOptionText(){
        return (TextView) LayoutInflater.from(toolbar.getContext())
                .inflate(R.layout.base__toolbar_option_text, toolbar, false);
    }

    public SVGImageView inflateOptionImage(){
        return (SVGImageView) LayoutInflater.from(toolbar.getContext())
                .inflate(R.layout.base__toolbar_option_svg, toolbar, false);
    }

    public View addOption(View view) {
        ((LinearLayout) toolbar.findViewById(R.id.base__toolbar_parent))
                .addView(view);

        return view;
    }

    public View addTitle(View view) {
        ((LinearLayout) toolbar.findViewById(R.id.base__toolbar_title_parent))
                .addView(view);

        return view;
    }

    public LinearLayout getParent(){
        return (LinearLayout) toolbar.findViewById(R.id.base__toolbar_parent);
    }

    public ToolbarLayoutBuilder setTransparentBackground() {
        getParent().setBackgroundColor(Color.TRANSPARENT);

        return this;
    }

    public ToolbarLayoutBuilder hide(){
        getParent().setVisibility(View.GONE);

        return this;
    }

}

File: src/main/java/com/guardanis/blank/tools/views/ToolbarLayoutBuilder.kt
package com.guardanis.blank.tools.views

import android.graphics.Color
import android.support.v7.app.ActionBar
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView

import com.guardanis.blank.R
import com.guardanis.imageloader.views.SVGImageView

class ToolbarLayoutBuilder {

    val toolbar: Toolbar

    val parent: LinearLayout
        get() = toolbar.findViewById<LinearLayout>(R.id.base__toolbar_parent)

    @JvmOverloads
    constructor(toolbar: Toolbar, inflatedLayoutResId: Int = R.layout.base__toolbar) {
        this.toolbar = toolbar

        val customView = LayoutInflater.from(toolbar.context)
                .inflate(inflatedLayoutResId, toolbar, false)

        toolbar.removeAllViews()
        toolbar.addView(customView)

        try {
            toolbar.setPadding(0, 0, 0, 0) // Fix for tablets
            toolbar.setContentInsetsAbsolute(0, 0)
        }
        catch (e: Throwable) { e.printStackTrace() }
    }

    fun addTitle(textResId: Int, onClickListener: (() -> Unit)?): ToolbarLayoutBuilder {
        return addTitle(toolbar.context.getString(textResId), onClickListener)
    }

    fun addTitle(text: String, onClickListener: (() -> Unit)?): ToolbarLayoutBuilder {
        val textTitle = inflateTitleText()

        textTitle.text = text
        textTitle.setOnClickListener({ onClickListener?.invoke() })

        addTitle(textTitle)

        return this
    }

    fun addTitleSvg(assetFileName: String, onClickListener: (() -> Unit)?): ToolbarLayoutBuilder {
        val svg = buildSvgImageView(assetFileName, onClickListener)
        addTitle(svg)

        return this
    }

    @JvmOverloads
    fun addOptionText(text: String, onClickListener: (() -> Unit)? = null): ToolbarLayoutBuilder {
        val textOption = buildTextOption(text, onClickListener)

        addOption(textOption)

        return this
    }

    fun addOptionSvg(assetFileName: String, onClickListener: (() -> Unit)?): ToolbarLayoutBuilder {
        val svg = buildSvgImageView(assetFileName, onClickListener)
        addOption(svg)

        return this
    }

    fun addOptionImage(resId: Int, onClickListener: (() -> Unit)?): ToolbarLayoutBuilder {
        val svg = inflateOptionImage()

        svg.setImageResource(resId)
        svg.setOnClickListener({ onClickListener?.invoke() })

        addOption(svg)

        return this
    }

    fun buildSvgImageView(assetFileName: String, onClickListener: (() -> Unit)?): SVGImageView {
        val svg = inflateOptionImage()
        svg.setImageAsset(assetFileName)
        svg.setOnClickListener({ onClickListener?.invoke() })

        return svg
    }

    @JvmOverloads
    fun buildTextOption(text: String, onClickListener: (() -> Unit)? = null): TextView {
        val textView = inflateOptionText()
        textView.text = text
        textView.setOnClickListener({ onClickListener?.invoke() })

        if (onClickListener != null)
            ViewHelper.setBackgroundResourceAndKeepPadding(textView, R.drawable.base__button_transparent)

        return textView
    }

    fun inflateTitleText(): TextView {
        return LayoutInflater.from(toolbar.context)
                .inflate(R.layout.base__toolbar_title_text, toolbar, false) as TextView
    }

    fun inflateOptionText(): TextView {
        return LayoutInflater.from(toolbar.context)
                .inflate(R.layout.base__toolbar_option_text, toolbar, false) as TextView
    }

    fun inflateOptionImage(): SVGImageView {
        return LayoutInflater.from(toolbar.context)
                .inflate(R.layout.base__toolbar_option_svg, toolbar, false) as SVGImageView
    }

    fun addOption(view: View): View {
        parent.addView(view)

        return view
    }

    fun addTitle(view: View): View {
        toolbar.findViewById<LinearLayout>(R.id.base__toolbar_title_parent)
                .addView(view)

        return view
    }

    fun setTransparentBackground(): ToolbarLayoutBuilder {
        return setBackgroundColor(Color.TRANSPARENT)
    }

    fun setBackgroundColor(color: Int): ToolbarLayoutBuilder {
        parent.setBackgroundColor(color)

        return this
    }

    fun setBackgroundResource(resourceId: Int): ToolbarLayoutBuilder {
        parent.setBackgroundResource(resourceId)

        return this
    }

    fun hide(): ToolbarLayoutBuilder {
        parent.visibility = View.GONE

        return this
    }

}
